awx/api/views/__init__.py

```
@@ -3826,6 +3826,168 @@ def get_queryset(self):
+    counter 2 is event 1, and has 6 children
+    etc.
+
+    The UI also needs to be able to collapse over "meta" events -- events that
+    show up due to verbosity or warnings from the system while the play is running.
+    These events have a 0 level event, with no parent uuid.
+
+        playbook_on_start
+        verbose
+          playbook_on_play_start
+            playbook_on_task_start
+              runner_on_start        <- level 3
+        verbose                      <- jump to level 0
+        verbose
+                runner_on_ok         <- jump back to level 3
+              playbook_on_task_start
+                runner_on_start
+                runner_on_ok
+        verbose
+        verbose
+              playbook_on_stats
+
+    These verbose statements that fall in the middle of a series of children events
+    are problematic for the UI.
+    To help, this view will attempt to place the events into the hierarchy, without
+    the event level jumps.
+
+        playbook_on_start
+          verbose
+          playbook_on_play_start
+            playbook_on_task_start
+              runner_on_start        <- A
+                verbose              <- this maps to the uuid of A
+                verbose
+                runner_on_ok
+              playbook_on_task_start <- B
+                runner_on_start
+                runner_on_ok
+                verbose              <- this maps to the uuid of B
+                verbose
+              playbook_on_stats
+
+    The output will include a JSON object where the key is the event counter,
+    and the value is the assigned nested uuid.
+    '''
+
+    renderer_classes = [JSONRenderer]
+    meta_events = ('debug', 'verbose', 'warning', 'error', 'system_warning', 'deprecated')
+
+    def get(self, request, **kwargs):
+        job = get_object_or_404(models.Job, pk=kwargs['pk'])
+        if not job.event_processing_finished:
+            return Response({})
```

If this won't run when event processing isn't finished, then why do we need to bother with the rowNumber? It should be the same as counter minus 1.

---

@AlanCoding does that mean there are never gaps in the counter values once processing is done? If that's the case, we can sort that out in the UI.

It also means I could probably look for ways to simplify the UI logic, as it does extra work to handle cases where there's a counter gap. Can there still be a gap while the job is running/processing?

---

There can't reasonably be a gap when `job.event_processing_finished` is True. Consider the logic from that.

https://github.com/ansible/awx/blob/6c1adade250020fcf900ecfc6d33e6f8274c8e0a/awx/main/models/unified_jobs.py#L1022

The way this works:
 - Control process receives output from receptor, and it does minimal processing of this and submits the data to receptor
   - This process counts the number of events that it sends, and saves that to `emitted_events`
 - Callback receiver pulls the data off the redis queue and then saves it to database

If `event_processing_finished` never transitions to `True`, then that's an API bug. Doesn't matter what type of job or what it does. This is accounting - the callback receiver should always save the same number of events that the control process submitted.

If the job is actually _running_ then this endpoint returns `{}` anyway

---

there are 3 numbers to consider here

- event_qs.count()
- emitted_events
- the `counter` field on the event itself (from ansible-runner)

the first 2 are pretty much guaranteed to match if event_processing_finished==true, as Alan outlined above.

What about `counter` field on the event itself? if there are 100 events, are we guaranteed that we have events with counters 1 through 100? Looking at code I don't see why not. I'm wondering if the previous reports of missing counters was because of not checking whether events had been fully processed or not.

We could make the decision to rely on `counter`, simplifying this endpoint a bit. We could just throw an error or warning if this code finds some missing counters.